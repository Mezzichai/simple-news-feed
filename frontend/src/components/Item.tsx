
type Story = {
  objectID: number;
  url: string;
  title: string;
  author: string;
  num_comments: number;
  points: number;
};


type ItemProps = {
  item: Story;
  onRemoveHandler: (id: number) => void;
};



const Item:React.FC<ItemProps> = ({ item, onRemoveHandler }) => (
  <li>
    <span>
      <a href={item.url}>{item.title}</a>
    </span>
    <span>{item.author}</span>
    <span>{item.num_comments}</span>
    <span>{item.points}</span>
    <button type="button" onClick={() => onRemoveHandler(item.objectID)}>
      Remove item
    </button>
  </li>
);

export default Item
