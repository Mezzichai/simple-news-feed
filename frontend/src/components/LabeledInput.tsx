import appStyles from '../styles/appStyles.module.css'

type LabeledInputProps = {
  id: string,
  onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  value: string
  type?: string
  isFocused: boolean
  children: React.ReactNode
};

const LabeledInput: React.FC<LabeledInputProps> = ({
  id, 
  onInputChange, 
  value, 
  type = "text",
  isFocused,
  children
}) => {
  return (
    //using fragments allow for "siblingization" without unnecessary praent elements
    <>
      <label htmlFor={id}>{children}</label>
      <input 
        className={appStyles.searchInput}
        id={id} 
        type={type}
        value={value} 
        autoFocus={isFocused}
        onChange={onInputChange}
      />
    </>
  );
}

export default LabeledInput