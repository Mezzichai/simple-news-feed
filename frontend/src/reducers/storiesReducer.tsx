
type Story = {
  objectID: number;
  url: string;
  title: string;
  author: string;
  num_comments: number;
  points: number;
};

type Stories = Story[];
  
type StoriesFetchInitAction = {
  type: 'STORIES_FETCH_INIT';
};

type StoriesFetchSuccessAction = {
  type: 'STORIES_FETCH_SUCCESS';
  payload: Stories;
};

type StoriesFetchFailureAction = {
  type: 'STORIES_FETCH_FAILURE';
};

type MoreStoriesFetchInitAction = {
  type: 'MORE_STORIES_FETCH_INIT';
};

type MoreStoriesFetchSuccessAction = {
  type: 'MORE_STORIES_FETCH_SUCCESS';
  payload: Stories;
};

type MoreStoriesFetchFailureAction = {
  type: 'MORE_STORIES_FETCH_FAILURE';
};


export type StoriesRemoveAction = {
  type: 'REMOVE_STORY';
  payload: number
};

export type SortOrder = "descending" | "ascending"

export type SortKey = 
  "title" | 
  "author" | 
  "num_comments" | 
  "points";

export type StoriesSortAction = {
  type: 'SORT_STORIES';
  payload: {
    sortOrder: SortOrder,
    sortKey: SortKey
  }
}

type StoriesAction =
  | StoriesFetchInitAction
  | StoriesFetchSuccessAction
  | StoriesFetchFailureAction
  | MoreStoriesFetchInitAction
  | MoreStoriesFetchSuccessAction
  | MoreStoriesFetchFailureAction
  | StoriesRemoveAction
  | StoriesSortAction;


type StoriesState = {
  data: Stories;
  isLoading: boolean;
  isError: boolean;
  isMoreLoading: boolean;
  isMoreError: boolean;
}

export const REMOVE_STORY = "REMOVE_STORY";
export const STORIES_FETCH_INIT = "STORIES_FETCH_INIT";
export const STORIES_FETCH_SUCCESS = "STORIES_FETCH_SUCCESS";
export const STORIES_FETCH_FAILURE = "STORIES_FETCH_FAILURE";

export const MORE_STORIES_FETCH_INIT = "MORE_STORIES_FETCH_INIT";
export const MORE_STORIES_FETCH_SUCCESS = "MORE_STORIES_FETCH_SUCCESS";
export const MORE_STORIES_FETCH_FAILURE = "MORE_STORIES_FETCH_FAILURE";

export const SORT_STORIES = "SORT_STORIES";

export const storiesReducer = (state: StoriesState, action: StoriesAction) => {
  switch (action.type) {
    case STORIES_FETCH_INIT:
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case STORIES_FETCH_SUCCESS:
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case STORIES_FETCH_FAILURE:
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case MORE_STORIES_FETCH_INIT:
      return {
        ...state,
        isMoreLoading: true,
        isMoreError: false,
      };
    case MORE_STORIES_FETCH_SUCCESS:
      return {
        ...state,
        isMoreLoading: false,
        isMoreError: false,
        data: action.payload,
      };
    case MORE_STORIES_FETCH_FAILURE:
      return {
        ...state,
        isMoreLoading: false,
        isMoreError: true,
      };
    case REMOVE_STORY:
      return {
        ...state,
        data: state.data.filter(
          (story: Story) => action.payload !== story.objectID
        )
      };
    case SORT_STORIES:
      return {
        ...state,
        data: state.data.sort((a, b) => {
          if (action.payload.sortOrder === 'descending') {
            if (a[action.payload.sortKey] < b[action.payload.sortKey]) {
              return -1
            } else if (a[action.payload.sortKey] > b[action.payload.sortKey]) {
              return 1
            } else {
              return 0
            }
          } else {
            if (a[action.payload.sortKey] < b[action.payload.sortKey]) {
              return 1
            } else if (a[action.payload.sortKey] > b[action.payload.sortKey]) {
              return -1
            } else {
              return 0
            }
          }
        })
      }
    default:
      throw new Error();
  }
};