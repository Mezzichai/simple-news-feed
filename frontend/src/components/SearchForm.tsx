import LabeledInput from "./LabeledInput"

type SearchFormProps = {
  searchTerm: string,
  handleSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void,
  handleSearchSubmit: (event: React.FormEvent<HTMLFormElement>) => void
}

const SearchForm:React.FC<SearchFormProps> = ({searchTerm, handleSearchInput, handleSearchSubmit}) => {
  return (
    <form onSubmit={handleSearchSubmit}>
      <LabeledInput
        id="search"
        value={searchTerm}
        isFocused
        onInputChange={handleSearchInput}
      >
        <strong>Search:</strong>
      </LabeledInput>
      <button type="submit" disabled={!searchTerm}>
          Submit
      </button>
  </form>
  )
}

export default SearchForm