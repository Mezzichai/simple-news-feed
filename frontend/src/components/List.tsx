import { memo, useState } from "react";
import Item from "./Item";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faChevronDown } from "@fortawesome/free-solid-svg-icons/faChevronDown";
import SortDropDown from "./SortDropDown";
import { SortKey, SortOrder } from "../reducers/storiesReducer";
import dropDownStyles from '../styles/dropDownStyles.module.css'
import SortOrderDropDown from "./SortOrderDropDown";

type Story = {
  objectID: number;
  url: string;
  title: string;
  author: string;
  num_comments: number;
  points: number;
};

type Stories = Story[];

type ListProps = {
  list: Stories;
  onRemoveHandler: (id: number) => void;
  handleChangeSortKey: (key: SortKey) => void;
  handleChangeSortOrder: (order: SortOrder) => void;
  sortKey: SortKey;
  sortOrder: SortOrder;
}

const propertyList = {
  title: {name: "title"}, 
  author: {name: "author"},
  num_comments: {name: "comments"}, 
  points: {name: "points"}
}

const orderList: SortOrder[] = [
  "ascending", "descending"
]

const List: React.FC<ListProps> =  memo(({ list, onRemoveHandler, handleChangeSortKey, handleChangeSortOrder, sortKey, sortOrder }) => {
  const [sortOptionsOpen, setSortOptionsList] = useState<boolean>(false);
  const [orderOptionsOpen, setOrderOptionsList] = useState<boolean>(false);

  return (
    <>
      <div className={dropDownStyles.sortContainers}>
        <div className={dropDownStyles.sortMenu}>
          <span style={{marginRight: "7px"}}>
            Sort by: {propertyList[sortKey].name}
          </span>
          <FontAwesomeIcon icon={faChevronDown} onClick={() => setSortOptionsList(!sortOptionsOpen)}/>
          {sortOptionsOpen ? <SortDropDown list={propertyList} handleClick={handleChangeSortKey} /> : null}
        </div>

        <div style={{marginLeft: "20px"}} className={dropDownStyles.sortMenu}>
          <span style={{marginRight: "7px"}}>
            Order: {sortOrder}
          </span>
          <FontAwesomeIcon icon={faChevronDown} onClick={() => setOrderOptionsList(!orderOptionsOpen)}/>
          {orderOptionsOpen ? <SortOrderDropDown list={orderList} handleClick={handleChangeSortOrder} /> : null}
        </div>
      </div>
      <ul>
        {list.map((item) => (
          <Item 
            key={item.objectID}
            item={item}
            onRemoveHandler={onRemoveHandler}
          />
        ))}
      </ul>
    </>
  );
})

export default List