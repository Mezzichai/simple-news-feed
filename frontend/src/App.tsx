import { useCallback, useEffect, useReducer, useState } from 'react';
import List from './components/List';
import useStorageState from './hooks/useStorageState';
import axios from 'axios';
import SearchForm from './components/SearchForm';
import { SortKey, REMOVE_STORY, STORIES_FETCH_FAILURE, STORIES_FETCH_INIT, STORIES_FETCH_SUCCESS, storiesReducer, SORT_STORIES, SortOrder, MORE_STORIES_FETCH_INIT, MORE_STORIES_FETCH_SUCCESS, MORE_STORIES_FETCH_FAILURE } from './reducers/storiesReducer';
import SearchHistory from './components/SearchHistory';
import appStyles from './styles/appStyles.module.css'

const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?";
const API_QUERY_PARAM = "query=";
const API_PAGE_PARAM = "&page=";


const getUrl = (searchTerm: string) => `${API_ENDPOINT}${API_QUERY_PARAM}${searchTerm}`;

const App = () => {
  
  const [searchTerm, setSearchTerm] = useStorageState(
    "search",
    "React"
  );

  const [history, setHistory] = useState<string[]>([getUrl(searchTerm)])

  const searchHistoryTerms = history.map(url => url.replace(`${API_ENDPOINT}${API_QUERY_PARAM}`, ""))

  const handleSearchInput = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  };

  const handleSearch = (searchTerm: string) => {
    const url = getUrl(searchTerm);
    setHistory(prevState => {
      const newHistory = [url, ...prevState];
      if (newHistory.length > 6) {
        newHistory.pop();
      }
      return newHistory;
    });
  };

  const handleSearchSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    handleSearch(searchTerm)
    event.preventDefault();
  };

  const [stories, dispatchStories] = useReducer(
    storiesReducer,
    { data: [], isLoading: false, isError: false, isMoreLoading: false, isMoreError: false}
  );

  const handleFetchStories = useCallback(() => {
    dispatchStories({ type: STORIES_FETCH_INIT });
    axios
      .get(history[0])
      .then(result => { 
        dispatchStories({
          payload: result.data.hits, 
          type: STORIES_FETCH_SUCCESS
        })
      })
      .catch(() => dispatchStories({ type: STORIES_FETCH_FAILURE })
      )
  }, [history]);

  useEffect(() => {
   handleFetchStories();
  }, [handleFetchStories]);

  const [page, setPage] = useState<number>(0)
  
  const handleFetchMoreStories = () => {
    dispatchStories({ type: MORE_STORIES_FETCH_INIT });
    axios
      .get(`${history[0]}${API_PAGE_PARAM}${page + 1}`)
      .then(result => { 
        dispatchStories({
          payload: stories.data.concat(result.data.hits), 
          type: MORE_STORIES_FETCH_SUCCESS
        })
        setPage(prevPage => prevPage + 1)
      })
      .catch(() => dispatchStories({ type: MORE_STORIES_FETCH_FAILURE })
      )
  }
  

  const handleRemoveStory = useCallback((id: number) => {
    dispatchStories({
      type: REMOVE_STORY,
      payload: id,
    });
  }, []);

 
  const [sortKey, setSortKey] = useState<SortKey>("title");
  const [sortOrder, setSortOrder] = useState<SortOrder>("descending");

  useEffect(() => {
    dispatchStories({
      type: SORT_STORIES,
      payload: {sortKey, sortOrder}
    })
  }, [sortKey, sortOrder]);

  const handleChangeSortKey = (key: SortKey) => {
    setSortKey(key)
  };

  const handleChangeSortOrder = (order: SortOrder) => {
    setSortOrder(order)
  };
  
  return (
    <div className={appStyles.container}>
      <h1>My Hacker Stories</h1>
      <SearchForm  
        searchTerm={searchTerm}
        handleSearchInput={handleSearchInput}
        handleSearchSubmit={handleSearchSubmit}
      />
      <SearchHistory history={searchHistoryTerms.slice(1)} handleSearch={handleSearch} />

      <hr />
      {stories.isError ?
        <p className={appStyles.statusInfo}>Error</p>
      : stories.isLoading ?
        <p className={appStyles.statusInfo}>Loading...</p>
      : 
        <>
          <List list={stories.data} onRemoveHandler={handleRemoveStory} handleChangeSortOrder={handleChangeSortOrder} handleChangeSortKey={handleChangeSortKey} sortKey={sortKey} sortOrder={sortOrder}/>
          {stories.isMoreError ?
            <>
              <p className={appStyles.statusInfo}>Error, try again</p>
              <button onClick={handleFetchMoreStories}>More</button>
            </>
          : stories.isMoreLoading ? 
            <p className={appStyles.statusInfo}>Loading...</p>
          :
            <button onClick={handleFetchMoreStories}>More</button>
          }
        </>
      }
    </div>
  );
};


export default App;