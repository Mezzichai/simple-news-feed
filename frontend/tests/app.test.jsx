import { describe, it, vi, expect } from 'vitest';
import Item from '../src/components/Item';
import SearchForm from '../src/components/SearchForm';
import App from '../src/App';
import LabeledInput from '../src/components/LabeledInput';
import List from '../src/components/List';
import axios from 'axios';
import { storiesReducer, REMOVE_STORY } from '../src/reducers/storiesReducer';
import {
  render,
  screen,
  fireEvent,
  waitFor,
  } from '@testing-library/react';
  

const storyOne = {
  title: 'React',
  url: 'https://reactjs.org/',
  author: 'Jordan Walke',
  num_comments: 3,
  points: 4,
  objectID: 0,
};

const storyTwo = {
  title: 'Redux',
  url: 'https://redux.js.org/',
  author: 'Dan Abramov, Andrew Clark',
  num_comments: 2,
  points: 5,
  objectID: 1,
};

const stories = [storyOne, storyTwo];

vi.mock('axios');

describe('App', () => {
  it('succeeds fetching data', async () => {
    const promise = Promise.resolve({
      data: {
        hits: stories,
      },
    });
    axios.get.mockImplementationOnce(() => promise);

    render(<App />);

    expect(screen.queryByText(/Loading/)).toBeInTheDocument();

    await waitFor(async () => await promise);

    expect(screen.queryByText(/Loading/)).toBeNull();

    expect(screen.getByText('React')).toBeInTheDocument();

    expect(screen.getByText('Redux')).toBeInTheDocument();

    expect(screen.getAllByText('Remove item').length).toBe(2);
  });


  it("fails fetching data", async () => {
    const promise = Promise.reject();

    axios.get.mockImplementationOnce(() => promise);

    render(<App />);

    expect(screen.queryByText(/Loading/)).toBeInTheDocument();

    try {
      await waitFor(async () => await promise);
    } catch (error) {
      expect(screen.queryByText(/Loading/)).toBeNull();
      expect(screen.queryByText(/Error/)).toBeInTheDocument();
    }
  });
  

  it("removes a story", async () => {
    const promise = Promise.resolve({
      data: {
        hits: stories,
      },
    });
    axios.get.mockImplementationOnce(() => promise);

    render(<App />);

    await waitFor(async () => await promise)

    //refers to the number of times this was found 
    expect(screen.getAllByText("Remove item").length).toBe(2);
    expect(screen.getByText('Jordan Walke')).toBeInTheDocument();

    fireEvent.click(screen.getAllByText("Remove item")[0]);

    expect(screen.getAllByText("Remove item").length).toBe(1);
    expect(screen.queryByText("Jordan Walke")).toBeNull()
  });

  it('searches for specific stories', async () => {
    const reactPromise = Promise.resolve({
      data: {
        hits: stories,
      },
    });

    const anotherStory = {
      title: 'JavaScript',
      url: 'https://en.wikipedia.org/wiki/JavaScript',
      author: 'Brendan Eich',
      num_comments: 15,
      points: 10,
      objectID: 3,
    };

    const javascriptPromise = Promise.resolve({
      data: {
        hits: [anotherStory],
      },
    });

    axios.get.mockImplementation((url) => {
      if (url.includes('React')) {
        return reactPromise;
      }

      if (url.includes('JavaScript')) {
        return javascriptPromise;
      }

      throw Error();
    });

    // Initial Render

    render(<App />);

    // First Data Fetching

    await waitFor(async () => await reactPromise);

    expect(screen.queryByDisplayValue('React')).toBeInTheDocument();
    expect(screen.queryByDisplayValue('JavaScript')).toBeNull();

    expect(screen.queryByText('Jordan Walke')).toBeInTheDocument();
    expect(
      screen.queryByText('Dan Abramov, Andrew Clark')
    ).toBeInTheDocument();
    expect(screen.queryByText('Brendan Eich')).toBeNull();

    // User Interaction -> Search

    fireEvent.change(screen.queryByDisplayValue('React'), {
      target: {
        value: 'JavaScript',
      },
    });

    expect(screen.queryByDisplayValue('React')).toBeNull();
    expect(
      screen.queryByDisplayValue('JavaScript')
    ).toBeInTheDocument();

    fireEvent.submit(screen.queryByText('Submit'));

    // Second Data Fetching

    await waitFor(async () => await javascriptPromise);

    expect(screen.queryByText('Jordan Walke')).toBeNull();
    expect(
      screen.queryByText('Dan Abramov, Andrew Clark')
    ).toBeNull();
    expect(screen.queryByText('Brendan Eich')).toBeInTheDocument();
  });

  it("renders snapshot", () => {
    const { container } = render(<App />);
    expect(container.firstChild).toMatchSnapshot();
  })
});

describe('stories reducer', () => {
  it('removes a storty from stories', () => {
    const action = { type: REMOVE_STORY, payload: storyOne.objectID};
    const state = { data: stories, isLoading: false, isError: false };

    const newState = storiesReducer(state, action)
    const expectedState = {
      data: [storyTwo],
      isLoading: false,
      isError: false,
    }
    expect(newState).toStrictEqual(expectedState)
  });
});


describe('Item', () => { 
  it("renders all properties", () => {
    render(<Item item={storyOne} />);

    expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
    expect(screen.getByText('React')).toHaveAttribute(
      'href',
      'https://reactjs.org/'
    );
  });

  it('renders a clickable dismiss button', () => {
    render(<Item item={storyOne} />);
    
    expect(screen.getByRole('button')).toBeInTheDocument();
  });
  
  it('dismiss button calls the callback handler on click', () => {
    const handleRemoveItem = vi.fn();

    render(<Item item={storyOne} onRemoveHandler={handleRemoveItem}/>);
    
    fireEvent.click(screen.getByRole("button"));

    expect(handleRemoveItem).toHaveBeenCalledTimes(1)
  });

  it("renders snapshot", () => {
    const { container } = render(<Item item={storyOne} />);
    expect(container.firstChild).toMatchSnapshot();
  })
});



describe('SearchForm', () => {
  const searchFormProps = {
    searchTerm: 'React',
    handleSearchInput: vi.fn(),
    handleSearchSubmit: vi.fn(),
  };

  it('renders the input field with its value', () => {
    render(<SearchForm {...searchFormProps} />);
    expect(screen.getByDisplayValue("React")).toBeInTheDocument()
  });

  it('renders the correct label', () => {
    render(<SearchForm {...searchFormProps} />);
    expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
  });

  it("calls handleSearchInput on input changes", () => {
    render(<SearchForm {...searchFormProps} />);

    fireEvent.change(screen.getByDisplayValue("React"), {
      target: { value: "Redux"},
    });
    expect(searchFormProps.handleSearchInput).toHaveBeenCalledTimes(1)

  });

  it("calls handleSearchSubmit on submit click", () => {
    render(<SearchForm {...searchFormProps} />);

    fireEvent.submit(screen.getByRole("button"));

    expect(searchFormProps.handleSearchSubmit).toHaveBeenCalledTimes(1);
  })

  it("renders snapshot", () => {
    const { container } = render(<SearchForm {...searchFormProps} />);
    expect(container.firstChild).toMatchSnapshot();
  })
});


// describe('List', () => {
//   const List = {
//     searchTerm: 'React',
//     handleSearchInput: vi.fn(),
//     handleSearchSubmit: vi.fn(),
//   };
// })