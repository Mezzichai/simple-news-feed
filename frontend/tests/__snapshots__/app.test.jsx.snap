// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders snapshot 1`] = `
<div>
  <h1>
    My Hacker Stories
  </h1>
  <form>
    <label
      for="search"
    >
      <strong>
        Search:
      </strong>
    </label>
    <input
      id="search"
      type="text"
      value="JavaScript"
    />
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
  <hr />
  <span>
    Loading...
  </span>
</div>
`;

exports[`Item > renders snapshot 1`] = `
<li>
  <span>
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span>
    Jordan Walke
  </span>
  <span>
    3
  </span>
  <span>
    4
  </span>
  <button
    type="button"
  >
    Remove item
  </button>
</li>
`;

exports[`SearchForm > renders snapshot 1`] = `
<form>
  <label
    for="search"
  >
    <strong>
      Search:
    </strong>
  </label>
  <input
    id="search"
    type="text"
    value="React"
  />
  <button
    type="submit"
  >
    Submit
  </button>
</form>
`;
